File: src/main/java/mnm/mods/tabbychat/client/gui/ChatBox.java
package mnm.mods.tabbychat.client.gui;

import com.google.common.collect.ImmutableSet;
import mnm.mods.tabbychat.TabbyChat;
import mnm.mods.tabbychat.client.AbstractChannel;
import mnm.mods.tabbychat.client.ChatManager;
import mnm.mods.tabbychat.client.DefaultChannel;
import mnm.mods.tabbychat.client.TabbyChatClient;
import mnm.mods.tabbychat.api.Channel;
import mnm.mods.tabbychat.api.ChannelStatus;
import mnm.mods.tabbychat.api.events.MessageAddedToChannelEvent;
import mnm.mods.tabbychat.client.UserChannel;
import mnm.mods.tabbychat.client.settings.ServerSettings;
import mnm.mods.tabbychat.client.settings.TabbySettings;
import mnm.mods.tabbychat.client.util.ScaledDimension;
import mnm.mods.tabbychat.util.ILocation;
import mnm.mods.tabbychat.util.Location;
import mnm.mods.tabbychat.util.Vec2i;
import mnm.mods.tabbychat.client.gui.component.BorderLayout;
import mnm.mods.tabbychat.client.gui.component.GuiPanel;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.FontRenderer;
import net.minecraft.client.gui.IGuiEventListener;
import net.minecraft.client.gui.screen.ChatScreen;
import net.minecraft.client.gui.screen.Screen;
import net.minecraft.client.renderer.Rectangle2d;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.math.MathHelper;
import net.minecraft.util.text.ITextComponent;
import net.minecraftforge.common.MinecraftForge;

import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.function.Predicate;

public class ChatBox extends GuiPanel {

    public static final ResourceLocation GUI_LOCATION = new ResourceLocation(TabbyChat.MODID, "textures/chatbox.png");

    private static ChatBox instance;

    private ChatArea chatArea;
    private ChatTray pnlTray;
    private TextBox txtChatInput;

    private boolean dragMode;
    private Vec2i drag;
    private Location tempbox;

    private List<AbstractChannel> channels = new ArrayList<>();
    private AbstractChannel active = DefaultChannel.INSTANCE;
    private Map<Channel, ChannelStatus> channelStatus = new HashMap<>();

    private ChatScreen chat;

    public ChatBox(TabbySettings settings) {
        super(new BorderLayout());
        instance = this;
        this.addComponent(pnlTray = new ChatTray(), BorderLayout.Position.NORTH);
        this.addComponent(chatArea = new ChatArea(), BorderLayout.Position.CENTER);
        this.addComponent(txtChatInput = new TextBox(), BorderLayout.Position.SOUTH);
        this.addComponent(new Scrollbar(chatArea), BorderLayout.Position.EAST);

        super.setLocation(settings.advanced.getChatboxLocation());

        this.channels.add(DefaultChannel.INSTANCE);
        this.pnlTray.addChannel(DefaultChannel.INSTANCE);

        this.setStatus(DefaultChannel.INSTANCE, ChannelStatus.ACTIVE);

        super.tick();

        MinecraftForge.EVENT_BUS.addListener(this::messageScroller);
        MinecraftForge.EVENT_BUS.addListener(this::addChatMessage);
    }

    public static ChatBox getInstance() {
        return instance;
    }

    public void update(ChatScreen chat) {
        this.chat = chat;
        if (chat.field_195139_w/*suggestions*/ != null && !(chat.field_195139_w/*suggestions*/.field_198505_b instanceof TCRect)) {
            chat.field_195139_w/*suggestions*/.field_198505_b = new TCRect(chat.field_195139_w/*suggestions*/.field_198505_b);
        }
    }

    private void messageScroller(MessageAddedToChannelEvent.Post event) {

        // compensate scrolling
        ChatArea chatbox = getChatArea();
        if (getActiveChannel() == event.getChannel() && chatbox.getScrollPos() > 0 && event.getId() == 0) {
            chatbox.scroll(1);
        }
    }

    private void addChatMessage(MessageAddedToChannelEvent.Post event) {
        AbstractChannel channel = (AbstractChannel) event.getChannel();
        addChannel(channel);
        setStatus(channel, ChannelStatus.UNREAD);
    }

    public void addChannels(Collection<AbstractChannel> active) {
        active.forEach(this::addChannel);
    }

    public Set<AbstractChannel> getChannels() {
        return ImmutableSet.copyOf(this.channels);
    }

    private void addChannel(AbstractChannel channel) {
        if (!this.channels.contains(channel)) {
            this.channels.add(channel);
            pnlTray.addChannel(channel);
            ChatManager.instance().save();
        }

    }

    public void removeChannel(AbstractChannel channel) {
        if (channels.contains(channel) && channel != DefaultChannel.INSTANCE) {
            channels.remove(channel);
            pnlTray.removeChannel(channel);
        }
        if (getActiveChannel() == channel) {
            setActiveChannel(DefaultChannel.INSTANCE);
        }
        ChatManager.instance().save();
    }

    @Nullable
    ChannelStatus getStatus(Channel chan) {
        return channelStatus.get(chan);
    }

    public void setStatus(AbstractChannel chan, @Nullable ChannelStatus status) {
        this.channelStatus.compute(chan, (key, old) -> {
            if (status == null || old == null || status.ordinal() < old.ordinal()) {
                return status;
            }
            return old;
        });
        if (status == ChannelStatus.ACTIVE) {
            chatArea.setChannel(chan);
        }
    }

    public void clearMessages() {
        this.channels.removeIf(Predicate.isEqual(DefaultChannel.INSTANCE).negate());

        this.pnlTray.clear();
        setStatus(DefaultChannel.INSTANCE, ChannelStatus.ACTIVE);
    }


    public AbstractChannel getActiveChannel() {
        return active;
    }

    public void setActiveChannel(AbstractChannel channel) {
        TextBox text = this.txtChatInput;


        if (active.isPrefixHidden()
                ? text.getText().trim().isEmpty()
                : text.getText().trim().equals(active.getPrefix())) {
            // text is the prefix, so remove it.
            text.setText("");
            if (!channel.isPrefixHidden() && !channel.getPrefix().isEmpty()) {
                // target has prefix visible
                text.getTextField().getTextField().setText(channel.getPrefix() + " ");
            }
        }
        // set max text length
        boolean hidden = channel.isPrefixHidden();
        int prefLength = hidden ? channel.getPrefix().length() + 1 : 0;

        text.getTextField().getTextField().setMaxStringLength(ChatManager.MAX_CHAT_LENGTH - prefLength);

        // reset scroll
        // TODO per-channel scroll settings?
        if (channel != active) {
            getChatArea().resetScroll();
        }
        setStatus(active, null);
        active = channel;
        setStatus(active, ChannelStatus.ACTIVE);

        runActivationCommand(channel);

    }

    private ServerSettings server() {
        return TabbyChatClient.getInstance().getServerSettings();
    }

    private void runActivationCommand(AbstractChannel channel) {
        String cmd = channel.getCommand();
        if (cmd.isEmpty()) {


            String pat;
            if (channel instanceof UserChannel) {
                pat = server().general.messageCommand.get();
            } else {
                pat = server().general.channelCommand.get();
            }
            if (pat.isEmpty()) {
                return;
            }
            String name = channel.getName();
            if (channel == DefaultChannel.INSTANCE) {
                name = server().general.defaultChannel.get();
            }
            // insert the channel name
            cmd = pat.replace("{}", name);

        }
        if (cmd.startsWith("/")) {
            if (cmd.length() > ChatManager.MAX_CHAT_LENGTH) {
                cmd = cmd.substring(0, ChatManager.MAX_CHAT_LENGTH);
            }
            Minecraft.getInstance().player.sendChatMessage(cmd);
        }
    }

    @Override
    public void render(int mouseX, int mouseY, float parTicks) {
        super.render(mouseX, mouseY, parTicks);
        if (mc.field_71456_v/*ingameGUI*/.getChatGUI().getChatOpen() && chat != null) {
            FontRenderer fr = Minecraft.getInstance().fontRenderer;
            ILocation loc = getLocation();
            final int height = fr.FONT_HEIGHT + 3;
            final int xPos = chat.commandUsagePosition + loc.getXPos();
            int yPos = loc.getYHeight() - chat.commandUsage.size() * height;
            if (chat.field_195139_w/*suggestions*/ != null) {
                chat.field_195139_w/*suggestions*/.render(mouseX, mouseY);
            } else if (xPos + chat.commandUsageWidth > loc.getXWidth()) {
                int i = 0;

                for(String s : chat.commandUsage) {
                    fill(0, chat.height - 14 - 12 * i, chat.commandUsageWidth + 1, chat.height - 2 - 12 * i, 0xff000000);
                    fr.drawStringWithShadow(s, 1, chat.height - 14 + 2 - 12 * i, -1);
                    ++i;
                }
            } else {
                for (String s : chat.commandUsage) {
                    fill(xPos - 1, yPos, xPos + chat.commandUsageWidth + 1, yPos - height, 0xd0000000);
                    fr.drawStringWithShadow(s, xPos, yPos - height + 2, -1);
                    yPos += height;
                }
            }

            ITextComponent itextcomponent = this.mc.field_71456_v/*ingameGUI*/.getChatGUI().getTextComponent((double) mouseX, (double) mouseY);
            if (itextcomponent != null && itextcomponent.getStyle().getHoverEvent() != null) {
                chat.renderComponentHoverEffect(itextcomponent, mouseX, mouseY);
            }
        }
    }

    @Override
    public boolean mouseClicked(double mouseX, double mouseY, int button) {
        if (button == 0 && (pnlTray.getLocation().contains(mouseX, mouseY)
                || Screen.hasAltDown() && getLocation().contains(mouseX, mouseY))) {
            dragMode = !pnlTray.isHandleHovered(mouseX, mouseY);
            drag = new Vec2i((int) mouseX, (int) mouseY);
            tempbox = getLocation().copy();
        }
        return super.mouseClicked(mouseX, mouseY, button);
    }

    @Override
    public boolean mouseDragged(double mx, double my, int mb, double mxd, double myd) {
        if (drag != null) {

            if (!dragMode) {
                setLocation(new Location(
                        tempbox.getXPos(),
                        tempbox.getYPos() + (int) my - drag.y,
                        tempbox.getWidth() + (int) mx - drag.x,
                        tempbox.getHeight() - (int) my + drag.y));
                ChatManager.instance().markDirty(active);
            } else {
                setLocation(getLocation().copy()
                        .setXPos(tempbox.getXPos() + (int) mx - drag.x)
                        .setYPos(tempbox.getYPos() + (int) my - drag.y));
            }
        }
        return super.mouseDragged(mx, my, mb, mxd, myd);
    }

    @Override
    public boolean mouseReleased(double mouseX, double mouseY, int button) {
        if (drag != null) {
            drag = null;
            tempbox = null;
        }
        return super.mouseReleased(mouseX, mouseY, button);
    }

    @Override
    public boolean mouseScrolled(double p_mouseScrolled_1_, double p_mouseScrolled_3_, double p_mouseScrolled_5_) {
        return this.chatArea.mouseScrolled(p_mouseScrolled_1_, p_mouseScrolled_3_, p_mouseScrolled_5_);
    }

    @Override
    public void tick() {
        ILocation bounds = getLocation();

        double scale = mc.gameSettings.chatScale;

        // original dims
        final int x = (int) (bounds.getXPos() * scale);
        final int y = (int) (bounds.getYPos() * scale);
        final int w = (int) (bounds.getWidth() * scale);
        final int h = (int) (bounds.getHeight() * scale);

        // the new dims
        int w1 = w;
        int h1 = h;
        int x1 = x;
        int y1 = y;

        final int SCREEN_W = mc.mainWindow.getScaledWidth();
        final int SCREEN_H = mc.mainWindow.getScaledHeight();

        // limits for sizes
        // FIXME 500 and 400 max is because of texture limit
        final int MIN_W = 50;
        final int MIN_H = 50;
        final int MAX_W = Math.min(500, SCREEN_W);
        final int MAX_H = Math.min(400, SCREEN_H);

        final int HOTBAR = 25;

        // calculate width and height first
        // used to calculate max x and y
        w1 = Math.max(MIN_W, w1);
        w1 = Math.min(MAX_W, w1);
        // this is different because height anchor is at the top
        // so is affected at the bottom.
        if (h1 < MIN_H) {
            y1 -= MIN_H - h1;
            h1 = MIN_H;
        }
        if (h1 > MAX_H) {
            y1 += h1 - MAX_H;
            h1 = MAX_H;
        }

        // limits for position
        final int MIN_X = 0;
        final int MIN_Y = 0;
        final int MAX_X = SCREEN_W - w1;
        final int MAX_Y = SCREEN_H - h1 - HOTBAR;

        // calculate x and y coordinates
        x1 = Math.max(MIN_X, x1);
        x1 = Math.min(MAX_X, x1);
        y1 = Math.max(MIN_Y, y1);
        y1 = Math.min(MAX_Y, y1);

        // reset the location if it changed.
        if (x1 != x || y1 != y || w1 != w || h1 != h) {
            setLocation(new Location(
                    MathHelper.ceil(x1 / scale),
                    MathHelper.ceil(y1 / scale),
                    MathHelper.ceil(w1 / scale),
                    MathHelper.ceil(h1 / scale)));
        }
        super.tick();
    }

    @Override
    public void setLocation(ILocation location) {
        super.setLocation(location);
        // save bounds
        TabbySettings sett = TabbyChatClient.getInstance().getSettings();
        sett.advanced.chatX.set(location.getXPos());
        sett.advanced.chatY.set(location.getYPos());
        sett.advanced.chatW.set(location.getWidth());
        sett.advanced.chatH.set(location.getHeight());
        sett.save();
    }

    @Override
    public void onClosed() {
        super.onClosed();
        tick();
    }

    @Nullable
    @Override
    public IGuiEventListener getFocused() {
        return txtChatInput;
    }

    public ChatArea getChatArea() {
        return this.chatArea;
    }

    public ChatTray getTray() {
        return this.pnlTray;
    }

    public TextBox getChatInput() {
        return this.txtChatInput;
    }

    public void onScreenHeightResize(int oldWidth, int oldHeight, int newWidth, int newHeight) {

        if (oldWidth == 0 || oldHeight == 0)
            return; // first time!

        // measure the distance from the bottom, then subtract from new height

        ScaledDimension oldDim = new ScaledDimension(oldWidth, oldHeight);
        ScaledDimension newDim = new ScaledDimension(newWidth, newHeight);

        int bottom = oldDim.getScaledHeight() - getLocation().getYPos();
        int y = newDim.getScaledHeight() - bottom;
        this.setLocation(getLocation().copy().setYPos(y));
        this.tick();
    }

    private class TCRect extends Rectangle2d {

        private final Rectangle2d parent;

        public TCRect(Rectangle2d parent) {
            super(0, 0, 0, 0);
            this.parent = parent;
        }

        @Override
        public int getX() {
            return Math.max(0, Math.min(parent.getX() + getLocation().getXPos(), chat.width - parent.getWidth()));
        }

        @Override
        public int getY() {
            return getLocation().getYHeight() - parent.getHeight() - 14 * getChatInput().getWrappedLines().size();
        }

        @Override
        public int getWidth() {
            return parent.getWidth();
        }

        @Override
        public int getHeight() {
            return parent.getHeight();
        }

        @Override
        public boolean contains(int x, int y) {
            return x >= this.getX() && x <= this.getX() + this.getWidth() && y >= this.getY() && y <= this.getY() + this.getHeight();
        }
    }

}

File: src/main/kotlin/mnm/mods/tabbychat/client/gui/ChatBox.kt
package mnm.mods.tabbychat.client.gui

import com.google.common.collect.ImmutableSet
import mnm.mods.tabbychat.MODID
import mnm.mods.tabbychat.api.Channel
import mnm.mods.tabbychat.api.ChannelStatus
import mnm.mods.tabbychat.api.events.MessageAddedToChannelEvent
import mnm.mods.tabbychat.client.ChannelImpl
import mnm.mods.tabbychat.client.ChatManager
import mnm.mods.tabbychat.client.DefaultChannel
import mnm.mods.tabbychat.client.TabbyChatClient
import mnm.mods.tabbychat.client.gui.widget.MultiLineTextFieldWidget
import mnm.mods.tabbychat.client.util.ScaledDimension
import mnm.mods.tabbychat.util.mc
import net.minecraft.client.gui.FocusableGui
import net.minecraft.client.gui.IGuiEventListener
import net.minecraft.client.gui.IRenderable
import net.minecraft.client.gui.screen.ChatScreen
import net.minecraft.client.gui.screen.Screen
import net.minecraft.client.renderer.Rectangle2d
import net.minecraft.util.ResourceLocation
import net.minecraft.util.math.MathHelper.ceil
import net.minecraft.util.math.MathHelper.floor
import org.lwjgl.glfw.GLFW
import thedarkcolour.kotlinforforge.forge.FORGE_BUS

object ChatBox : FocusableGui(), IRenderable, ILocatable {

    override var xPos: Int = 0
    override var yPos: Int = 0
    override var width: Int = 0
    override var height: Int = 0

    val GUI_LOCATION = ResourceLocation(MODID, "textures/chatbox.png")
    var chatArea: ChatArea
        private set
    var tray: ChatTray
        private set
    var chatInput: MultiLineTextFieldWidget
        private set
    var scrollbar: Scrollbar

    private val channels = ArrayList<Channel>()
    private val children = ArrayList<IGuiEventListener>()

    class ChannelStatusMap {
        private val channelStatus = mutableMapOf<Channel, ChannelStatus>()
        internal operator fun get(chan: Channel): ChannelStatus? {
            return channelStatus[chan]
        }

        operator fun set(chan: Channel, status: ChannelStatus?) {
            channelStatus.compute(chan) { _, old ->
                if (status?.ordinal ?: -1 < old?.ordinal ?: ChannelStatus.values().size + 1) {
                    status
                } else {
                    old
                }
            }
            if (status == ChannelStatus.ACTIVE) {
                chatArea.channel = chan
            }
        }
    }

    var status = ChannelStatusMap()

    var activeChannel: Channel = DefaultChannel
        set(channel) {
            val text = this.chatInput

            val prefix = text.text.trim { it <= ' ' }
            val oldChannel = field
            if (oldChannel is ChannelImpl) {
                if (oldChannel.isPrefixHidden && prefix.isEmpty() || prefix == oldChannel.prefix) {
                    // text is the prefix, so remove it.
                    text.text = ""
                    if (channel is ChannelImpl) {
                        if (!channel.isPrefixHidden && channel.prefix.isNotEmpty()) {
                            // target has prefix visible
                            text.text = channel.prefix + " "
                        }
                    }
                }
            }

            if (channel is ChannelImpl) {
                // set max text length
                val hidden = channel.isPrefixHidden
                val prefLength = if (hidden) channel.prefix.length + 1 else 0
                text.maxStringLength = ChatManager.MAX_CHAT_LENGTH - prefLength
            } else {
                text.maxStringLength = ChatManager.MAX_CHAT_LENGTH
            }

            // reset scroll
            // TODO per-channel scroll settings?
            if (channel !== field) {
                chatArea.resetScroll()
            }
            status[field] = null
            field = channel
            status[field] = ChannelStatus.ACTIVE
        }

    private lateinit var chat: ChatScreen

    private var locationDirty = false

    private fun saveLocation() {
        val loc = TabbyChatClient.settings.advanced.chatLocation
        loc.xPos = xPos
        loc.yPos = yPos
        loc.width = width
        loc.height = height

        TabbyChatClient.settings.save()
        locationDirty = true
    }

    init {
        val location = TabbyChatClient.settings.advanced.chatLocation
        xPos = location.xPos
        yPos = location.yPos
        width = location.width
        height = location.height

        tray = ChatTray(this)
        chatArea = ChatArea(this)
        chatInput = MultiLineTextFieldWidget(this, mc.fontRenderer, "")
        scrollbar = Scrollbar(chatArea)

        this.children += listOf<IGuiEventListener>(tray, chatArea, chatInput)

        this.channels.add(DefaultChannel)
        this.tray.addChannel(DefaultChannel)

        status[DefaultChannel] = ChannelStatus.ACTIVE

        FORGE_BUS.addListener(::messageScroller)
        FORGE_BUS.addListener(::addChatMessage)
    }

    override fun children(): MutableList<out IGuiEventListener> {
        return children
    }

    fun init(screen: Screen) {
        this.chat = screen as ChatScreen
        this.chat.commandSuggestionHelper.field_228095_d_ = chatInput
        val chan = activeChannel
        if (screen.defaultInputFieldText.isEmpty()
                && chan is ChannelImpl
                && !chan.isPrefixHidden
                && chan.prefix.isNotEmpty()) {
            screen.defaultInputFieldText = chan.prefix + " "
        }
        screen.inputField = chatInput
        chatInput.text = screen.defaultInputFieldText

        chatInput.setTextFormatter(screen.commandSuggestionHelper::func_228122_a_)
        chatInput.setResponder(screen::func_212997_a)

        val children = screen.children() as MutableList<IGuiEventListener>
        children[0] = chatInput
    }

    fun tick() {
        // try to fix the chatbox location to keep it in the window.
        if (normalizeLocation()) {
            ChatManager.markDirty(activeChannel)
            this.locationDirty = true
        }

        tray.tick()
        chatInput.tick()

        // finally, save the location if it changed.
        if (!draggingChatBox && locationDirty) {
            saveLocation()
            locationDirty = false
        }
    }

    private fun update() {
        chat.commandSuggestionHelper.suggestions?.apply {
            if (field_228138_b_ !is TCRect) {
                field_228138_b_ = TCRect(field_228138_b_)
            }
        }
    }

    private fun messageScroller(event: MessageAddedToChannelEvent.Post) {

        // compensate scrolling
        val chatbox = chatArea
        if (activeChannel === event.channel && chatbox.scrollPos > 0 && event.id == 0) {
            chatbox.scroll(1)
        }
    }

    private fun addChatMessage(event: MessageAddedToChannelEvent.Post) {
        val channel = event.channel
        addChannel(channel)
        status[channel] = ChannelStatus.UNREAD
    }

    fun addChannels(active: Collection<Channel>) {
        active.forEach { this.addChannel(it) }
    }

    fun getChannels(): Set<Channel> {
        return ImmutableSet.copyOf(this.channels)
    }

    private fun addChannel(channel: Channel) {
        if (!this.channels.contains(channel)) {
            this.channels.add(channel)
            tray.addChannel(channel)
            ChatManager.save()
        }

    }

    fun removeChannel(channel: Channel) {
        if (channels.contains(channel) && channel !== DefaultChannel) {
            channels.remove(channel)
            tray.removeChannel(channel)
        }
        if (activeChannel === channel) {
            activeChannel = DefaultChannel
        }
        ChatManager.save()
    }

    fun clearMessages() {
        this.channels.removeIf { it !== DefaultChannel }

        this.tray.clearMessages()
        status[DefaultChannel] = ChannelStatus.ACTIVE
    }

    /*
    private val settings get() = TabbyChatClient.serverSettings

    private fun runActivationCommand(channel: ChannelImpl) {
        var cmd = channel.command
        if (cmd.isEmpty()) {
            val pat = if (channel is UserChannel) {
                settings.general.messageCommand
            } else {
                settings.general.channelCommand
            }
            if (pat.isEmpty()) {
                return
            }
            var name = channel.name
            if (channel === DefaultChannel) {
                name = settings.general.defaultChannel
            }
            // insert the channel name
            cmd = pat.replace("{}", name)

        }
        if (cmd.startsWith("/")) {
            if (cmd.length > ChatManager.MAX_CHAT_LENGTH) {
                cmd = cmd.substring(0, ChatManager.MAX_CHAT_LENGTH)
            }
            mc.player?.sendChatMessage(cmd)
        }
    }

     */

    override fun render(x: Int, y: Int, parTicks: Float) {
        handleDragging(x.toDouble(), y.toDouble())

        this.chatArea.render(x, y, parTicks)

        if (mc.ingameGUI.chatGUI.chatOpen) {
            this.tray.render(x, y, parTicks)
            this.chatInput.render(x, y, parTicks)
            update()
            chat.commandSuggestionHelper.render(x, y)

            val itextcomponent = mc.ingameGUI.chatGUI.getTextComponent(x.toDouble(), y.toDouble())
            if (itextcomponent != null && itextcomponent.style.hoverEvent != null) {
                chat.renderComponentHoverEffect(itextcomponent, x, y)
            }
        }
    }

    override fun keyPressed(keyCode: Int, scanCode: Int, modifiers: Int): Boolean {
        if (keyCode == GLFW.GLFW_KEY_ENTER || keyCode == GLFW.GLFW_KEY_KP_ENTER) {
            mc.ingameGUI.chatGUI.resetScroll()
            chat.sendMessage(chatInput.text)
            chatInput.text = chat.defaultInputFieldText

            if (!TabbyChatClient.settings.advanced.keepChatOpen) {
                mc.displayGuiScreen(null)
            }
            return true
        }

        if (chat.commandSuggestionHelper.onKeyPressed(keyCode, scanCode, modifiers)) {
            return true
        }
        if (keyCode == GLFW.GLFW_KEY_TAB) {
            return true
        }
        return super.keyPressed(keyCode, scanCode, modifiers)
    }

    private var draggingChatBox = false
    private var resizing = false

    private var dragStartX = 0
    private var dragStartY = 0

    private var tempX = 0
    private var tempY = 0
    private var tempW = 0
    private var tempH = 0

    override fun mouseClicked(x: Double, y: Double, button: Int): Boolean {
        if (chat.commandSuggestionHelper.onClick(x, y, button)) {
            return true
        }
        if (button == 0 && tray.contains(x.toInt(), y.toInt())
                || Screen.hasAltDown() && this.contains(x.toInt(), y.toInt())) {
            draggingChatBox = true
            resizing = tray.isHandleHovered(x, y)
            dragStartX = x.toInt()
            dragStartY = y.toInt()
            tempX = this.xPos
            tempY = this.yPos
            tempW = this.width
            tempH = this.height
        }

        if (mc.ingameGUI.chatGUI.chatOpen) {
            val itextcomponent = mc.ingameGUI.chatGUI.getTextComponent(x, y)
            if (itextcomponent != null && itextcomponent.style.clickEvent != null) {
                return chat.handleComponentClicked(itextcomponent)
            }
        }

        return super.mouseClicked(x, y, button)
    }

    private fun handleDragging(mx: Double, my: Double) {
        if (draggingChatBox) {
            if (resizing) {
//                xPos = tempX + mx.toInt() - dragStartX
                yPos = tempY + my.toInt() - dragStartY
                width = tempW + mx.toInt() - dragStartX
                height = tempH - my.toInt() + dragStartY
            } else {
                xPos = tempX + mx.toInt() - dragStartX
                yPos = tempY - dragStartY + my.toInt()
            }
            if (normalizeLocation()) {
                ChatManager.markDirty(activeChannel)
            }
        }
    }

    override fun mouseReleased(x: Double, y: Double, button: Int): Boolean {
        if (draggingChatBox) {
            draggingChatBox = false
            locationDirty = true
        }
        return super.mouseReleased(x, y, button)
    }

    override fun mouseScrolled(x: Double, y: Double, scroll: Double): Boolean {
        return this.chat.commandSuggestionHelper.onScroll(scroll) || super.mouseScrolled(x, y, scroll)
    }

    /**
     * Normalizes the chatbox to keep it in view.
     */
    private fun normalizeLocation(): Boolean {
        val scale = mc.gameSettings.chatScale

        // original dims
        val x = floor(xPos * scale)
        val y = floor(yPos * scale)
        val w = floor(width * scale)
        val h = floor(height * scale)

        // the new dims
        var w1 = w
        var h1 = h
        var x1 = x
        var y1 = y

        val screenW = mc.mainWindow.scaledWidth
        val screenH = mc.mainWindow.scaledHeight

        val hotbar = 25

        // limits for sizes
        val minW = 150
        val minH = 50
        val maxH = screenH - hotbar


        // calculate width and height first
        // used to calculate max x and y
        w1 = w1.coerceIn(minW, screenW)

        // this is different because height anchor is at the top
        // so is affected at the bottom.
        if (h1 < minH) {
            y1 -= minH - h1
            h1 = minH
        }
        if (h1 > maxH) {
            y1 += h1 - maxH
            h1 = maxH
        }

        // limits for position
        val minX = 0
        val minY = 0
        val maxX = screenW - w1
        val maxY = screenH - h1 - hotbar

        // calculate x and y coordinates
        x1 = x1.coerceIn(minX, maxX)
        y1 = y1.coerceIn(minY, maxY)

        // reset the location if it changed.
        if (x1 != x || y1 != y || w1 != w || h1 != h) {
            this.xPos = ceil(x1 / scale)
            this.yPos = ceil(y1 / scale)
            this.width = ceil(w1 / scale)
            this.height = ceil(h1 / scale)
            return true
        }
        return false
    }

    override fun getFocused(): IGuiEventListener? {
        return chatInput
    }

    fun onScreenHeightResize(oldWidth: Int, oldHeight: Int, newWidth: Int, newHeight: Int) {

        if (oldWidth == 0 || oldHeight == 0)
            return  // first time!

        // measure the distance from the bottom, then subtract from new height

        val oldDim = ScaledDimension(oldWidth, oldHeight)
        val newDim = ScaledDimension(newWidth, newHeight)

        val bottom = oldDim.scaledHeight - yPos
        val y = newDim.scaledHeight - bottom
        if (this.yPos != y) {
            this.yPos = y
            locationDirty = true
        }
    }

    private class TCRect(private val parent: Rectangle2d) : Rectangle2d(0, 0, 0, 0) {

        override fun getX(): Int {
            return parent.x
        }

        override fun getY(): Int {
            chatInput.y
            return chatInput.y - parent.height + chatInput.wrappedLines.size * mc.fontRenderer.FONT_HEIGHT
        }

        override fun getWidth(): Int {
            return parent.width
        }

        override fun getHeight(): Int {
            return parent.height
        }

        override fun contains(x: Int, y: Int): Boolean {
            return this.x <= x && x <= this.x + this.width && this.y <= y && y <= this.y + this.height
        }
    }

}
